thentos-adhocracy/tests/Thentos/Adhocracy3/Backend/Api/SimpleSpec.hs

```diff
@@ -91,6 +95,40 @@ spec =
                      (Aeson.eitherDecode userdata :: Either String A3UserWithPass)
                      `shouldBe` Left "Illegal whitespace sequence in user name: \" Anna  Toll\""
 
+        describe "login" . before setupTestBackend . after teardownTestBackend $
+            it "rejects bad credentials mimicking A3" $
+                \bts -> runTestBackend bts $ do
+
+                    let reqBody = encodePretty . object $
+                            [ "name"     .= String "Anna Müller"
+                            , "password" .= String "this-is-wrong"
+                            ]
+
+                    rsp <- srequest $ makeSRequest "POST" "login_username" [] reqBody
+                    shouldBeErr400WithCustomMessage (simpleStatus rsp) (simpleBody rsp)
+                        "\"User doesn't exist or password is wrong\""
+
+        describe "activate_account" . before setupTestBackend . after teardownTestBackend $
+            it "rejects bad path mimicking A3" $
+                \bts -> runTestBackend bts $ do
+
+                    let reqBody = encodePretty . object $
+                            [ "path" .= String "/activate/no-such-path" ]
+
+                    rsp <- srequest $ makeSRequest "POST" "activate_account" [] reqBody
```

Comment:
OK, good point about `[json|...]`. I still think we could use `shouldSatisfy` or whatever and actually test exactly what we want rather than just substring, since the specification we want is that it be a particular json format, *not* that it be valid json and contain certain substrings.

I'm trying to improve the speed of the test suite, and soon the overhead of setting up each test should be negligible (at least until we move to some other DB), and to moreover be able to run them in parallel. It is very hard currently to tell from our tests what is the specification we're trying to match against, and what just happened to be expedient (or, even more importantly, not expedient) to test. I really don't think we should be skimping on making our tests readable and complete.